zkevm-circuits/src/mpt_circuit/helpers.rs

```diff
@@ -0,0 +1,1382 @@
+                            key_data[is_s.idx()].drifted_num_nibbles.expr(),
+                            key_data[is_s.idx()].drifted_is_odd.expr(),
+                        );
+                        let key_rlc = key_rlc.expr() + config.drifted_rlp_key.key.expr(
+                            cb,
+                            config.drifted_rlp_key.key_value.clone(),
+                            key_mult.expr(),
+                            is_key_odd.expr(),
+                            r
+                        );
+                        // The key of the drifted leaf needs to match the key of the leaf
+                        require!(key_rlc => expected_key_rlc[is_s.idx()]);
+
+                        // Total number of nibbles needs to be KEY_LEN_IN_NIBBLES
+                        // (RLC encoding could be the same for addresses with zero's at the end)
+                        let num_nibbles = num_nibbles::expr(config.drifted_rlp_key.key_value.len(), is_key_odd.expr());
+                        require!(key_num_nibbles.expr() + num_nibbles => KEY_LEN_IN_NIBBLES);
+
+                        // Multiplier after list and key
+                        let mult = config.drifted_rlp_key.rlp_list.rlp_mult(r) * drifted_item.mult();
+
+                        // Complete the drifted leaf rlc by adding the bytes on the value row
+                        let leaf_rlc = (config.drifted_rlp_key.rlc(r), mult.expr()).rlc_chain(leaf_no_key_rlc[is_s.idx()].expr());
+                        // The drifted leaf needs to be stored in the branch at `drifted_index`.
+                        require!((1, leaf_rlc, config.drifted_rlp_key.rlp_list.num_bytes(), parent_data[is_s.idx()].drifted_parent_rlc.expr()) => @KECCAK);
+                    }
+                }}
+            }}
+            config
+        })
+    }
+
+    pub(crate) fn assign<S: ChallengeSet<F>>(
+        &self,
+        region: &mut CachedRegion<'_, '_, F, S>,
+        offset: usize,
+        parent_data: &[ParentDataWitness<F>],
+        drifted_list_bytes: &[u8],
+        drifted_item: &RLPItemWitness,
+        _r: F,
+    ) -> Result<(), Error> {
+        if parent_data[true.idx()].is_placeholder || parent_data[false.idx()].is_placeholder {
+            self.drifted_rlp_key
+                .assign(region, offset, drifted_list_bytes, drifted_item)?;
+        }
+        Ok(())
+    }
+}
+
+/// Handles wrong leaves
+#[derive(Clone, Debug, Default)]
+pub struct WrongGadget<F> {
+    wrong_rlp_key: ListKeyGadget<F>,
+    wrong_mult: Cell<F>,
+    pub(crate) is_key_equal: IsEqualGadget<F>,
+    wrong_key: Option<Expression<F>>,
+}
+
+impl<F: Field> WrongGadget<F> {
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn construct(
+        cb: &mut MPTConstraintBuilder<F>,
+        expected_address: Expression<F>,
+        is_non_existing: Expression<F>,
+        key_value: &RLPItemView<F>,
+        key_rlc: &Expression<F>,
+        wrong_item: &RLPItemView<F>,
```

Comment:
@Brechtpd Maybe `wrong_item` is a bit misleading as it actually contains the address of the enquired address? The "wrong" address is actually stored in the `key_rlc`.